cranelift/isle/docs/language-reference.md

```diff
@@ -949,6 +949,118 @@ semantically as important as the core language: they are not
 implementation details, but rather, a well-defined interface by which
 ISLE can interface with the outside world (an "FFI" of sorts).
 
+### If-Let Clauses
+
+As an extension to the basic left-hand-side / right-hand-side rule
+idiom, ISLE allows *if-let clauses* to be used. These add additional
+pattern-matching steps, and can be used to perform additional tests
+and also to use constructors in place of extractors during the match
+phase when this is more convenient.
+
+To introduce the concept, an example follows (this is taken from the
+[RFC](https://github.com/bytecodealliance/rfcs/tree/main/isle-extended-patterns.md)
+that proposed if-lets):
+
+```lisp
+;; `u32_fallible_add` can now be used in patterns in `if-let` clauses
+(decl pure u32_fallible_add (u32 u32) u32)
+(extern constructor u32_fallible_add u32_fallible_add)
+
+(rule (lower (load (iadd addr
+                         (iadd (uextend (iconst k1))
+                               (uextend (iconst k2))))))
+      (if-let k (u32_fallible_add k1 k2))
+      (isa_load (amode_reg_offset addr k)))
```

Comment:
It's a good question for sure. IMHO it gets sort of complicated if we take as granted that we don't want to change syntax for cases without if-let clauses... that implies that (if we want straightforward parsing without unlimited lookahead at least) cases *with* if-lets need a wrapper form around the pattern, like `(rule (let PAT (if-let ...) (if-let ...)) RHS)` maybe, reusing the `let` keyword. It could work; happy to discuss further and we can always refactor if we find an improved syntax :-)

Reply:
Yeah, that's basically what peepmatic used: `(=> PATTERN RHS)` and `(=> (when PATTERN PREDICATES...) RHS)`. I didn't love it but it did work.

Reply:
Hmm, yeah, it's not immediately clear to me that this is less confusing -- seeing the prefix/wrapper form first, if one doesn't know what it does, is possibly worse than seeing a top-level pattern that looks like the other rules and then additional conditions separately...

Regarding which part of the rule the `if-let` clauses are a part of: if I squint the right way, it seems like it actually kind of doesn't matter? In other words I can build a consistent mental model for what's going on as "go down the list of patterns/conditions, and when we reach the end we have the answer". That's a level of understanding that's more basic than the strict phase separation we talk about in the term-rewriting flow, but it's (IMHO) not too bad for a casual user tweaking some rules.

I think I'll err on the side of the syntax in the RFC and go ahead and merge this but absolutely we can still talk more about it if you (or anyone else) have further thoughts!